libjulia/interpreter/Interpreter.cpp

```diff
@@ -0,0 +1,225 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * IULIA interpreter.
+ */
+
+#include <libjulia/interpreter/Interpreter.h>
+
+#include <libjulia/interpreter/EVMInstructionInterpreter.h>
+
+#include <libsolidity/inlineasm/AsmData.h>
+
+#include <libsolidity/interface/Exceptions.h>
+
+#include <libdevcore/FixedHash.h>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace dev;
+using namespace dev::julia;
+
+
+void Interpreter::operator()(ExpressionStatement const& _expressionStatement)
+{
+	evaluateMulti(_expressionStatement.expression);
+}
+
+void Interpreter::operator()(Assignment const& _assignment)
+{
+	solAssert(_assignment.value, "");
+	vector<u256> values = evaluateMulti(*_assignment.value);
+	solAssert(values.size() == _assignment.variableNames.size(), "");
+	for (size_t i = 0; i < values.size(); ++i)
+	{
+		string const& varName = _assignment.variableNames.at(i).name;
+		solAssert(m_variables.count(varName), "");
+		m_variables[varName] = values.at(i);
+	}
+}
+
+void Interpreter::operator()(VariableDeclaration const& _declaration)
+{
+	vector<u256> values(_declaration.variables.size(), 0);
+	if (_declaration.value)
+		values = evaluateMulti(*_declaration.value);
+
+	solAssert(values.size() == _declaration.variables.size(), "");
+	for (size_t i = 0; i < values.size(); ++i)
+	{
+		string const& varName = _declaration.variables.at(i).name;
+		solAssert(!m_variables.count(varName), "");
+		m_variables[varName] = values.at(i);
+	}
+}
+
+void Interpreter::operator()(If const& _if)
+{
+	solAssert(_if.condition, "");
+	if (evaluate(*_if.condition) != 0)
+		(*this)(_if.body);
+}
+
+void Interpreter::operator()(Switch const& _switch)
+{
+	solAssert(_switch.expression, "");
+	u256 val = evaluate(*_switch.expression);
+	for (auto const& c: _switch.cases)
```

Comment:
Should have an assert for `_switch.cases.size() >= 1`.

Reply:
Actually, what is the point allowing a `switch` statement with only `default` case (after we introduced `if`)?